pyrokinetics/local_geometry/LocalGeometry.py

```diff
@@ -26,14 +109,425 @@ def __init__(self, *args, **kwargs):
             super(LocalGeometry, self).__init__(_data_dict)
 
     # TODO replace this with an abstract classmethod
-    @not_implemented
-    def load_from_eq(self, eq, psi_n=None):
+    def from_global_eq(
```

Comment:
I agree that these probably should be a  `classmethod` though I'm not 100% sure about how to implement it. I tried but did not succeed :( 

Where would you move them to?

Reply:
I wouldn't worry about it for this PR, since it would be most easily achieved in a more serious refactor and I think what you have here is good. In a future update, I'm imagining something like this:

```python
class LocalGeometryMiller(LocalGeometry):
    # Have __init__ take just Miller parameters and a few optional args
    def __init__(self, rho, Rmaj, zmid, shat, q, kappa, delta, zeta, s_kappa, s_delta, s_zeta, ntheta=64):
        self.rho = rho
        self.Rmaj = Rmaj
        ...
        self.s_zeta = s_zeta
        # Calculate self.R, self.Z, self.b_poloidal, self.bunit_over_b0, etc using Miller params
        # (This will use a bunch of overridden functions such as get_flux_surface, get_b_poloidal, etc)
        super().__init__(ntheta)


    # Provide alternative constructors
    @classmethod
    def from_global_eq(cls, eq, psi_n, ntheta=64):
        rho = eq.rho(psi_n)
        ...
        s_zeta = calculate_s_zeta_somehow()
        return cls(rho, Rmaj, zmid, shat, q, kappa, delta, zeta, s_kappa, s_delta, s_zeta, ntheta=ntheta)
```

We'd have to write a specialised `__init__` for each type of `LocalGeometry`, and then each `from_global_eq` would have to determine every parameter using `Equilibrium` data only and without touching any attributes of `LocalGeometry` itself.

The point about maybe moving this responsibility to `Equilibrium` is more of a philosophical issue than a practical one. Currently, each local geometry must have the ability to parse and understand the entire global equilibrium. I feel like it would be 'cleaner' for an `Equilibrium` to know how to produce subsets of itself than for the subsets to produce themselves from the whole, if that makes any sense. I made some steps towards it in PR https://github.com/pyro-kinetics/pyrokinetics/pull/128, where `Equilibrium` can produce a generic `FluxSurface` object, with the intention of `LocalGeometry` objects building themselves from that instead. From there, you could have something like:

```python
class Equilibrium:
    ...
    def local_geometry(self, psi_n, fit):
        # This would probably be defined elsewhere:
        fittings = {"Miller": LocalGeometryMiller, "FourierCGYRO": LocalGeometryFourierCGYRO, ...}
        fs = self.flux_surface(psi_n)
        return fittings[fit].from_flux_surface(fs)
```

The only real difference is that `LocalGeometry` no longer has any idea that `psi_n` or any wider equilibrium exists, regardless of whether it's created from an `Equilibrium`, a `GKInput`, or just user inputs. Practically speaking, it doesn't matter much either way.

Reply:
Hmm I see what you mean. 

A generic `FluxSurface` Object is essentially a `LocalGeometry` Object but without the shaping coefficients. Something could be done on this though we need to think about whether there is enough of a difference here. But philosophically I agree with you.